feat(color) - Use generic YAML parser to save/load Theme yaml files.

philmoz opened the pull request:
Summary of changes:
- Remove hard wired code and use YAML parser for theme files.
- Fix handling of empty input lines in YAML parser.
- Increase size of YAML parser variables to handle theme 'info' property length.

pfeerick commented on radio/src/gui/colorlcd/theme_manager.cpp:
```diff
@@ -21,27 +21,119 @@
 
-#define HEX_COLOR_VALUE_LEN 8
-constexpr const char *RGBSTRING = "RGB(";
 constexpr const char* SELECTED_THEME_FILE = THEMES_PATH "/selectedtheme.txt";
```
If there is any way you can see to finally getting rid of this also, and instead relying on the `themeName` parameter in `RADIO/radio.yml` that would be great. As it would finally bring all the configuration setting into yaml proper, and opening up avenues for future work. Both the radio firmware and companion look at that file to determine the active theme.

philmoz replied:
> If there is any way you can see to finally getting rid of this also, and instead relying on the `themeName` parameter in `RADIO/radio.yml` that would be great. As it would finally bring all the configuration setting into yaml proper, and opening up avenues for future work. Both the radio firmware and companion look at that file to determine the active theme.

The 'themeName' property in the radio.yml files appears to be meant for some other use related to the OpenTXTheme class. The radio writes an empty string to this field while Companion does not write it at all to radio.yml.

Can it be repurposed as the user selected theme (field size will need to be increased)?

pfeerick replied:
I can't actually see the purpose of that themeName property any more - this is all very old OTX -> ETX code, so has never been cleaned up properly. i.e. g_eeGeneral.themeName is only used in two places, and the radio actually just writes `themeName: ""` whether you are using the default theme or a custom theme... :shrug: I think it's also the reason we still have a `FlyskyTheme` knocking around...

philmoz commented:
I've moved the selected theme to the radio.yml file. I added a new property for this as I was not sure about re-using the existing themeName property.

One issue with this PR is that the YAML writer in the radio puts quotes around strings.
The old theme parser does not handle this so if you save a theme then revert to an old version of EdgeTX the theme details will have quotes in them on the radio. Not sure if this is worth trying to fix.

elecpower commented:
> One issue with this PR is that the YAML writer in the radio puts quotes around strings.

Quotes are good for the Companion yaml parser.

elecpower commented:
> One issue with this PR is that the YAML writer in the radio puts quotes around strings.
The old theme parser does not handle this so if you save a theme then revert to an old version of EdgeTX the theme details will have quotes in them on the radio. Not sure if this is worth trying to fix.

IMO not worth trying to fix. If you upgrade then revert with most apps be prepared for things to break when using the older version. That is why upgrade notes have a statement to backup before upgrading.

philmoz commented:
> > One issue with this PR is that the YAML writer in the radio puts quotes around strings.
> 
> Quotes are good for the Companion yaml parser.

Interesting you should say that because on my Mac, Companion discards all the quotes around strings when it saves the .ETX file.

elecpower commented:
> Interesting you should say that because on my Mac, Companion discards all the quotes around strings when it saves the .ETX file.

On encoding I believe it depends on whether the string contains 'special/problematic to yaml parser' characters as to whether a string is double quoted or not which is why yaml-cpp has had problems with some of the radio yaml encoder strings.
You can manually force double quotes on yaml-cpp encoding.

elecpower commented:
@raphaelcoeffic and to a lesser extend myself had many hours of 'fun' coercing the two yaml beasts to play nicely in the sandpit

ParkerEde commented:
I noticed in connection with this PR that from the selectedtheme.txt the entry:
`/THEMES/RK_Cappuccino/theme.yml`
is not correctly imported into the radio.yml. This causes the effect that after the next radio start the EdgeTx default theme is activated.

philmoz commented:
> I noticed in connection with this PR that from the selectedtheme.txt the entry:
> `/THEMES/RK_Cappuccino/theme.yml`
> is not correctly imported into the radio.yml. This causes the effect that after the next radio start the EdgeTx default theme is activated.

I tried with your files on my TX16S and it correctly selected the theme on restart.

Was the 'selectedtheme.txt' file deleted from your radio on restart?

ParkerEde commented:
Yes, the selectedtheme.txt was deleted after restart.
I got the issue with Companion Simulator with HorusX10 Express profile and with a real Horus X10 Express.

philmoz commented:
> Yes, the selectedtheme.txt was deleted after restart.
> I got the issue with Companion Simulator with HorusX10 Express profile and with a real Horus X10 Express.

Do you have an .ETX file that has the issue in the simulator?

ParkerEde commented:
Yes. Please remove .zip as filetype
[test.etx.zip](https://github.com/EdgeTX/edgetx/files/11230034/test.etx.zip)

ParkerEde commented:
I have now tested everything again more thoroughly with the main from today. With the firmware on the Horus X10 Express and the standalone firmware simulator, it runs correctly.
The only problem is that when the simulator is started from Companion, it deletes the selectedtheme.txt from the path of the SD card, but does not import this info into the radio.yml. This seems logical to me, because the simulator from Companion only accesses read-only. Therefore, I think it should not delete the selectedtheme.txt from the SD card path.

philmoz commented:
The first time you run the simulator from Companion it will work; but when you restart the simulator the theme reverts to default because the .ETX file is not updated by the simulator.

If you do the update on the radio and then import from the radio into companion it will have the theme stored correctly in the radio.yml file.

TBH, I'm not sure this is worth fixing.
